crates/task-impls/src/vote.rs

```diff
@@ -0,0 +1,514 @@
+        ViewSyncCommitCertificate2, ViewSyncFinalizeCertificate2, ViewSyncPreCommitCertificate2,
+    },
+    simple_vote::{
+        DAVote, QuorumVote, TimeoutVote, VIDVote, ViewSyncCommitVote, ViewSyncFinalizeVote,
+        ViewSyncPreCommitVote,
+    },
+    traits::{election::Membership, node_implementation::NodeType},
+    vote::{Certificate, HasViewNumber, Vote, VoteAccumulator},
+};
+use snafu::Snafu;
+use tracing::{debug, error};
+
+#[derive(Snafu, Debug)]
+/// Stub of a vote error
+pub struct VoteTaskError {}
+
+/// Task state for collecting votes of one type and emiting a certificate
+pub struct VoteCollectionTaskState<
+    TYPES: NodeType,
+    VOTE: Vote<TYPES>,
+    CERT: Certificate<TYPES, Voteable = VOTE::Commitment> + Debug,
+> {
+    /// Public key for this node.
+    pub public_key: TYPES::SignatureKey,
+
+    /// Membership for voting
+    pub membership: Arc<TYPES::Membership>,
+
+    /// accumulator handles aggregating the votes
+    pub accumulator: Option<VoteAccumulator<TYPES, VOTE, CERT>>,
+
+    /// The view which we are collecting votes for
+    pub view: TYPES::Time,
+
+    /// global event stream
+    pub event_stream: ChannelStream<HotShotEvent<TYPES>>,
+
+    /// Node id
+    pub id: u64,
+}
+
+/// Describes the functions a vote must implement for it to be aggregatable by the generic vote collection task
+pub trait AggregatableVote<
+    TYPES: NodeType,
+    VOTE: Vote<TYPES>,
+    CERT: Certificate<TYPES, Voteable = VOTE::Commitment>,
+>
+{
+    /// return the leader for this votes
+    fn get_leader(&self, membership: &TYPES::Membership) -> TYPES::SignatureKey;
+
+    /// return the Hotshot event for the completion of this CERT
+    fn make_cert_event(certificate: CERT, key: &TYPES::SignatureKey) -> HotShotEvent<TYPES>;
+}
+
+impl<
+        TYPES: NodeType,
+        VOTE: Vote<TYPES> + AggregatableVote<TYPES, VOTE, CERT>,
+        CERT: Certificate<TYPES, Voteable = VOTE::Commitment> + Debug,
+    > VoteCollectionTaskState<TYPES, VOTE, CERT>
+{
+    /// Take one vote and accumultate it. Returns either the cert or the updated state
+    /// after the vote is accumulated
+    pub async fn accumulate_vote(mut self, vote: &VOTE) -> (Option<HotShotTaskCompleted>, Self) {
+        if vote.get_leader(&self.membership) != self.public_key {
+            return (None, self);
+        }
+
+        if vote.get_view_number() != self.view {
+            error!(
+                "Vote view does not match! vote view is {} current view is {}",
+                *vote.get_view_number(),
+                *self.view
+            );
+            return (None, self);
+        }
+        let Some(accumulator) = self.accumulator else {
+            return (None, self);
+        };
+        match accumulator.accumulate(vote, &self.membership) {
+            Either::Left(acc) => {
+                self.accumulator = Some(acc);
+                (None, self)
+            }
+            Either::Right(qc) => {
```

Comment:
nit: I feel like we should use `cert` here instead of `qc` since we aren't always actually forming a quorum certificate.